### PR TITLE
Openconnect: T4656: Add listen-host field to specific the address to …

### PR DESCRIPTION
**Pull request by abxuz**
…listen on.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4656

## Component(s) name

Openconnect VPN
## Proposed changes

Add support to specific which address for the service to listen on
## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/vpn-openconnect.xml.in [hunk 1]
```diff
@@ -148,6 +148,32 @@
               </node>
             </children>
           </node>
+          <leafNode name="listen-host">
```
Comment (sever-sever) [inline interface-definitions/vpn-openconnect.xml.in]: Could you change it to `listen-address`?

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
Could we set several “listen-addresses”?
Should it listen by default and IPv6?

**Comment by abxuz** [comment]
> 

That might be unable to set  several “listen-addresses” 'cause the ocserv only supports one address to listen.

**Comment by sever-sever** [comment]
> > 
> 
> That might be unable to set several “listen-addresses” 'cause the ocserv only supports one address to listen.

The thing is, by default `ocserv` listen both IPv4 and IPv6 addresses
```
vyos@r14# sudo netstat -tulpn | grep 443
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      3407/ocserv-main    
tcp6       0      0 :::443                  :::*                    LISTEN      3407/ocserv-main    
udp        0      0 0.0.0.0:443             0.0.0.0:*                           3407/ocserv-main    
udp6       0      0 :::443                  :::*                                3407/ocserv-main  
```

**Comment by abxuz** [comment]
@sever-sever 
I manually edited the ocserv config file "/var/run/ocserv.conf" and added two lines of 'listen-host' trying to listen on two address. 
But finally i only got the last address listened.








And I seeked through the source code of ocserv, it turned out that the type of listen_host in config struct is "char *" which means it only stores one address.





As for the default behavior when no listen_host is configured, the ocserv uses "getaddrinfo" func to resolv the listen_host, the function may returns both ipv4 and ipv6 address (0.0.0.0 and ::)，so it listens on both ipv4 and ipv6 address.

**Review by sever-sever — APPROVED** [review]

File: interface-definitions/vpn-openconnect.xml.in [hunk 2]
```diff
@@ -148,6 +148,32 @@
               </node>
             </children>
           </node>
+          <leafNode name="listen-address">
```
Comment (c-po) [inline interface-definitions/vpn-openconnect.xml.in]: Please use our generic building block instead: `#include <include/listen-address.xml.i>`

We also do not listen on FQDNs ... we like IP addresses :)

**Review by c-po — CHANGES_REQUESTED** [review]

**Pull request merged**